torchqdynamics/utils.py

```diff
@@ -46,10 +46,28 @@ def ket_overlap(x: Tensor, y: Tensor) -> torch.complex:
     return (ket_to_bra(x) @ y).squeeze(-1).sum(-1)
 
 
-def ket_fidelity(x: Tensor, y: Tensor) -> float:
-    r"""Return the fidelity between two state vectors (kets).
+def sqrtm(x: Tensor) -> Tensor:
+    """Compute the square root of a symmetric or Hermitian positive definite matrix.
 
-    The fidelity between two pure states $\varphi$, $\psi$ is defined by their
+    Args:
+        x: Tensor of dimension `(..., n, n)`.
+
+    Returns:
+        Tensor of dimension `(..., n, n)` holding the square root of `x`.
+    """
+    # code copied from
+    # https://github.com/pytorch/pytorch/issues/25481#issuecomment-1032789228
+    L, Q = torch.linalg.eigh(x)
+    zero = torch.zeros((), device=L.device, dtype=L.dtype)
+    threshold = L.max(-1).values * L.size(-1) * torch.finfo(L.dtype).eps
+    L = L.where(L > threshold.unsqueeze(-1), zero)  # zero out small components
+    return (Q * L.sqrt().unsqueeze(-2)) @ Q.mH
```

Comment:
I don't know how important this is, but an sqrtm implementation based on `eigh` can be a bit slow, and also unstable numerically (probably why you have to zero out small components). Maybe a better implementation would be the following, based on a Newton method:
```python
class MatrixSquareRoot(Function):
    """Square root of a positive definite matrix.
    NOTE: matrix square root is not differentiable for matrices with
          zero eigenvalues.
    """    
    @staticmethod
    def forward(ctx, input):
        dim = input.shape[0]
        norm = torch.norm(input)
        Y = input/norm
        I = torch.eye(dim,dim,device=input.device).type(input.dtype)
        Z = torch.eye(dim,dim,device=input.device).type(input.dtype)
        for i in range(20):
            T = 0.5*(3.0*I - Z.mm(Y))
            Y = Y.mm(T)
            Z = T.mm(Z)
        sqrtm = Y*torch.sqrt(norm)
        ctx.mark_dirty(Y,I,Z)
        ctx.save_for_backward(sqrtm)
        return sqrtm

    @staticmethod
    def backward(ctx, grad_output):
        grad_input = None
        sqrtm, = ctx.saved_tensors
        dim = sqrtm.shape[0]
        norm = torch.norm(sqrtm)
        A = sqrtm/norm
        I = torch.eye(dim, dim, device=sqrtm.device).type(input.dtype)
        Q = grad_output/norm
        for i in range(20):
            Q = 0.5*(Q.mm(3.0*I-A.mm(A))-A.t().mm(A.t().mm(Q)-Q.mm(A)))
            A = 0.5*A.mm(3.0*I-A.mm(A))
        grad_input = 0.5*Q
        return grad_input    

symsqrt = MatrixSquareRoot.apply
```
This is faster, stable, and also supports backward differentiation. However, one has to manage the number of iteration of the process, which may be a bit tedious. Alternatively, we could support both implementations with a keyword argument in `sqrtm`. 

The code is copied from https://github.com/pytorch/pytorch/issues/25481#issuecomment-584896176.

Reply:
Nice! I'll let you add this in another PR 😉 (the proposed version supports backward differentiation with the addition of the incoming fix)